docs: Add snap upgrades how-to (#934)
canonical/k8s-snap: docs/src/snap/howto/upgrades.md

# Upgrading the snap

Upgrading the Kubernetes version of a node is a critical operation that 
requires careful planning and execution. {{product}} is shipped as a snap,
which simplifies the upgrade process.
This how-to guide will cover the steps to upgrade the {{product}} snap 
and how to put a hold on upgrades if desired.

## Important Considerations Before Upgrading

- According to the [upstream Kubernetes][1], skipping **minor** versions while 
upgrading is not supported. For more details, please visit the
[Version Skew Policy][2].
- Before performing an upgrade, it's important to back up the cluster data. 
This can be done by following the steps outlined in the [backup guide][3].
- For more information on managing snap updates, please refer to the 
[snap documentation][4].

## Patch Upgrade

Patch upgrades address bug fixes and are typically safe, introducing no 
breaking changes.
Snaps automatically check for updates of their specific track 

> (e.g. `1.32-classic`) several times a day and apply them when available.

Author: This is what I initially had, but changed according to this comment: https://github.com/canonical/k8s-snap/pull/934#discussion_r1908370666

> # Upgrading the snap

Reviewer: We need to get consensus on this capitalization on headers and sub-headers. @nhennigan what are we doing?

Upgrading the Kubernetes version of a node is a critical operation that 
requires careful planning and execution. {{product}} is shipped as a snap,
which simplifies the upgrade process.

> This how-to guide will cover the steps to upgrade the {{product}} snap to a 
> new version as well as how to freeze upgrades.

Reviewer: ```suggestion
This how-to guide will cover the steps to upgrade the {{product}} snap and how to put a
hold on upgrades if desired.
```

## Important Considerations Before Upgrading

- According to the [upstream Kubernetes][1], skipping **minor** versions while 
upgrading is not supported. For more details, please visit the
[Version Skew Policy][2].
- Before performing an upgrade, it's important to back up the cluster data. 
This can be done by following the steps outlined in the [backup guide][3].
- For more information on managing snap updates, please refer to the 
[snap documentation][4].

## Patch Upgrade

Patch upgrades address bug fixes and are typically safe, introducing no 
breaking changes.

> {{product}} snap is installed using a specific track (e.g. `1.32-classic`). 
> Snaps automatically check for updates several times a day and apply 
> them when available.
> These updates ensure the latest changes in the installed track are applied.

Reviewer: ```suggestion
Snaps automatically check for updates of their specific track (e.g. `1.32-classic`) several times
a day and apply them when available.
These updates ensure that the latest changes in the installed track are applied.
```
Let's make it flow a little better :)

Patch upgrades can also be triggered manually by following the steps below.

1. **List available revisions:**
```
snap info k8s
```

2. **Refresh the snap:**
```
snap refresh k8s
```

3. **Verify the upgrade:**
Ensure that the upgrade was successful by checking the version of the snap and 
confirming that the cluster is ready:
```
snap info k8s
sudo k8s status --wait-ready
```

## Minor Version Upgrade

Minor versions add new features or deprecate existing features without 
breaking changes.
To upgrade to a new minor version, the snap channel needs to be changed.

1. **List available channels:**
```
snap info k8s
```

2. **Change the snap channel:**
The {{product}} snap channel can be changed by using the `snap refresh` 
command.
```
snap refresh --channel=1.33/stable k8s
```

3. **Verify the upgrade:**
Ensure that the upgrade was successful by checking the version of the snap 
and confirming that the cluster is ready:
```
snap info k8s
sudo k8s status --wait-ready
```

```{note}
In a multi-node cluster, the upgrade should be performed on all nodes.
```

## Freezing Upgrades

To prevent automatic updates, the snap can be frozen to a specific revision. 

> `snap refresh --hold[=<duration>]` holds refreshes for a specified duration 

Reviewer: ```suggestion
`snap refresh --hold[=<duration>]` holds snap refreshes for a specified duration 
```

# Upgrading the snap

Upgrading the Kubernetes version of a node is a critical operation that 
requires careful planning and execution. {{product}} is shipped as a snap,
which simplifies the upgrade process.
This how-to guide will cover the steps to upgrade the {{product}} snap to a 
new version as well as how to freeze upgrades.

## Important Considerations Before Upgrading

- According to the [upstream Kubernetes][1], skipping **minor** versions while 

> upgrading is not supported. For more details, please visit [Version Skew Policy][2].

Reviewer: ```suggestion
upgrading is not supported. For more details, please visit the [Version Skew Policy][2].
```

- Before performing an upgrade, it's important to back up the cluster data. 
This can be done by following the steps outlined in the [backup guide][3].

> - For more info about managing snap updates, please refer to the [snap documentation][4].

Reviewer: ```suggestion
- For more information on managing snap updates, please refer to the [snap documentation][4].
```

> - Managing upgrades in a multi node cluster is a complex operation, and it's 

Reviewer: +1 let's remove this line

recommended to use lifecycle management tools like [Juju][5] or [Cluster API][6].

## What you will need

> - A running {{product}} cluster.

Reviewer: Let's remove this, it's a given.

## Patch Upgrade

> Patch upgrades apply bug fixes and are generally safe without API changes.

Reviewer 2: Suggestion 2: Patch upgrades address bug fixes and are typically safe, introducing no breaking changes.

> Snaps update automatically, and by default, the snapd daemon checks for updates 4 times a day.

Reviewer 2: This feels like an implemenatation detail of the Snap ecosystem.

Maybe:
```
Snaps automatically check for updates several times a day and apply them when available.
```

{{product}} is installed using a specific track, which consists of a major and minor version.
As an example, if {{product}} is installed using the `1.32` track, 
the snap updates automatically to the latest revision of `1.32` during daily checks.

> Patch upgrades can also be triggered manually by following the steps below.

Reviewer: Can we please condense this information to about half?

1. **List available revisions:**
```
snap info k8s
```

2. **Refresh the snap:**
```
snap refresh k8s
```

3. **Verify the upgrade:**
Ensure that the upgrade was successful by checking the version of the snap and 
confirming that the cluster is ready:
```
snap info k8s
k8s status --wait-ready
```

## Minor Version Upgrade

Minor versions add new features or deprecate existing features without breaking changes.
To upgrade to a new minor version, the snap channel needs to be changed.

1. **List available channels:**
```
snap info k8s
```

2. **Change the snap channel:**
The {{product}} snap channel can be changed either by using the `snap refresh` command or by using the `snap switch` command.

> The main difference between these two methods is that `snap refresh` will update the snap to the latest revision in the new channel, while `snap switch` switches to the new channel without doing a refresh.

Reviewer 3: I am a bit confused on the difference. Does this mean the snap switch will go to the oldest version of the new channel automatically?

Reviewer 2: snap switch will not actually update/refresh the snap but only change the track it follows. 

Think of it like changing the `origin` of a git repository (without actually fetching).

Reviewer 2: Maybe we should just explain one way of doing this (`snap refresh`) as `snap switch` is more of an edge case?

```
snap refresh --channel=1.33/stable k8s
```
Or
```
snap switch --channel=1.33/stable k8s
```

3. **Verify the upgrade:**
Ensure that the upgrade was successful by checking the version of the snap and 
confirming that the cluster is ready:
```
snap info k8s

> k8s status --wait-ready

Reviewer: Do we need sudo?

```

> 

Reviewer 3: Maybe add a line on how you need to do this for every snap node in the cluster

## Freezing Upgrades

To prevent automatic updates, the snap can be frozen to a specific revision. 
`snap refresh --hold[=<duration>]` holds refreshes for a specified duration (or forever, if no value is specified).
```
snap refresh k8s --hold
```

> Or

Reviewer: ```suggestion
Or specify a time window:
```
